pkg/awsclient/iam.go

```diff
@@ -0,0 +1,179 @@
+package awsclient
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/go-logr/logr"
+	awsv1alpha1 "github.com/openshift/aws-account-operator/pkg/apis/aws/v1alpha1"
+	"github.com/openshift/aws-account-operator/pkg/controller/utils"
+)
+
+func ListIAMUserTags(reqLogger logr.Logger, client Client, userName string) (*iam.ListUserTagsOutput, error) {
+	input := &iam.ListUserTagsInput{
+		UserName: aws.String(userName),
+	}
+
+	result, err := client.ListUserTags(input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeNoSuchEntityException:
+				fmt.Println(iam.ErrCodeNoSuchEntityException, aerr.Error())
+			case iam.ErrCodeServiceFailureException:
+				fmt.Println(iam.ErrCodeServiceFailureException, aerr.Error())
+			default:
+				fmt.Println(aerr.Error())
+			}
+		} else {
+			// Print the error, cast err to awserr.Error to get the Code and
+			// Message from an error.
+			fmt.Println(err.Error())
+		}
+		return result, err
+	}
+
+	return result, nil
+}
+
+func ListIAMUsers(reqLogger logr.Logger, client Client) ([]*iam.User, error) {
+	input := &iam.ListUsersInput{}
+	// List of IAM users to return
+	iamUserList := []*iam.User{}
+	pageNum := 0
+
+	err := client.ListUsersPages(input,
+		func(page *iam.ListUsersOutput, lastPage bool) bool {
+			iamUserList = append(iamUserList, page.Users...)
+			return pageNum <= 3
+		})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case iam.ErrCodeServiceFailureException:
+				msg := fmt.Sprintf("Service failure exception")
+				utils.LogAwsError(reqLogger, msg, nil, err)
+				return iamUserList, err
+			default:
+				msg := fmt.Sprintf("Unexpected AWS error")
+				utils.LogAwsError(reqLogger, msg, nil, err)
+				return iamUserList, err
+			}
+		} else {
+			// Print the error, cast err to awserr.Error to get the Code and
+			// Message from an error.
+			utils.LogAwsError(reqLogger, "Unexpected error when listing IAM users", nil, err)
+			fmt.Println(err.Error())
+		}
+		return iamUserList, err
+	}
+
+	return iamUserList, nil
+
+}
+
+// checkIAMUserExists checks if a given IAM user exists within an account
+// Takes a logger, an AWS client for the target account, and a target IAM username
+func CheckIAMUserExists(reqLogger logr.Logger, client Client, userName string) (bool, *iam.GetUserOutput, error) {
+	// Retry when getting IAM user information
+	// Sometimes we see a delay before credentials are ready to be user resulting in the AWS API returning 404's
+	var iamGetUserOutput *iam.GetUserOutput
+	var err error
+
+	attempt := 1
```

Comment:
Do we really need a second variable for this? It doesn't seem like `attempt` is used in any way that `i` couldn't be used.